peas/tests/serial_handlers/protocol_arduino_simulator.py

```diff
@@ -0,0 +1,422 @@
+
+from pocs.tests import serial_handlers
+import pocs.utils.logger
+
+
+def _drain_queue(q):
+    cmd = None
+    while not q.empty():
+        cmd = q.get_nowait()
+    return cmd  # Present just for debugging.
+
+
+class ArduinoSimulator:
+    """Simulates the serial behavior of the PANOPTES Arduino sketches.
+
+    The RS-232 connection is simulated with an input and output queue of bytes. This class provides
+    a run function which can be called from a Thread to execute. Every two seconds while running it
+    will generate another json output line,
+    at around 1000 bytes per second (near 9600 buad). The bytes  (actually chunked up in batches, so not emitting a single byte
+    every 1/960th of a second or so).
+    """
+
+    def __init__(self, relay_queue, json_queue, stop, logger):
+        """
+        Args:
+            relay_queue: The queue.Queue instance from which replay command bytes are read
+                and acted upon. Elements are of type bytes.
+            json_queue: The queue.Queue instance to which json messages (serialized to bytes)
+                are written at ~9600 baud. Elements are of type bytes.
+            stop: a threading.Event which is checked to see if run should stop executing.
+            logger: the Python logger to use for reporting messages.
+        """
+        self.relay_queue = relay_queue
+        self.json_queue = json_queue
+        self.stop = stop
+        self.logger = logger
+        # Time between producing messages.
+        self.message_delta = datetime.timedelta(seconds=2)
+        self.next_message_time = None
+        # Size of a chunk of bytes.
+        self.chunk_size = 20
+        # Interval between outputing chunks of bytes.
+        chunks_per_second = 1000.0 / self.chunk_size
+        chunk_interval = 1.0 / chunks_per_second
+        self.logger.debug('chunks_per_second={}   chunk_interval={}',
+                          chunks_per_second, chunk_interval)
+        self.chunk_delta = datetime.timedelta(seconds=chunk_interval)
+        self.next_chunk_time = None
+        self.pending_json_bytes = bytearray()
+        self.pending_relay_bytes = bytearray()
+        self.start_time = datetime.datetime.now()
+        self.report_num = 0
+        # TODO(jamessynge): Come up with a way to select which board is emulated. For now just
+        # one board is supported.
+        # TODO(jamessynge): Update camera_board message soon when camera board sketch is updated.
+        self.camera_board = json.loads("""
+            {
+                "name":"camera_board",
+                "millis":1507497677,
+                "report_num":7835,
+                "inputs":6,
+                "camera_00":1,
+                "camera_01":1,
+                "accelerometer": {"x":-7.02, "y":6.95, "z":1.70, "o": 6},
+                "humidity":59.60, "temp_00":12.50
+            }
+            """)
+        # TODO(jamessynge): Update telemetry_board message to reflect latest changes.
```

Comment:
We should update telemetry, camera, and power board to have consistent interface. For instance in telemetry the power is all under the `power` key but in camera it is just `camera_00`, which doesn't really explain what it is. Both of them need a version as well.

Reply:
Agreed, will file an issue.